### ISSUE TITLE
read_csv: support for compressed .zip files

### ISSUE THREAD

**Issue by raayu83** [pola-rs/polars#19447]
### Description

It seems read_csv does not yet support reading directly from .zip files.

It would be great if passing a zip file Path would try reading all files in the .zip into a df.
This is a pretty common usecase when dealing with reporting APIs.

Additionally, it would be great if there was a way to only read a single file in the zip. E.g. by passing a ZipFile Path, though I'd personally prefer an option to pass the filename directly without using ZipFile.

**Comment by mdavis-xyz** [comment]
I would love this too. My use case is reading Australian electricity market data ([example](https://www.nemweb.com.au/Data_Archive/Wholesale_Electricity/MMSDM/2025/MMSDM_2025_05/MMSDM_Historical_Data_SQLLoader/DATA/)). To convert to Parquet (for actual analysis), I need to unzip, then gzip (because the uncompressed file is too large for disk). Then point Polars at the gzipped file.

I think this is where the new decompression handler would need to go:

https://github.com/pola-rs/polars/blob/b9d05c48d58f611b19451ef5c6df60f55be6e444/crates/polars-io/src/utils/compression.rs#L38

**Comment by mdavis-xyz** [comment]
For my use case, I expect only one csv per zip. Some people may have multiple csvs per zip, where they want to select a specific CSV within the zip file. How should that be handled, in terms of the API? e.g. [duckdb-zipfs](https://github.com/isaacbrodsky/duckdb-zipfs) takes a string like `zip://examples/a.zip/a.csv`. Or should we add another argument to `scan_csv`. e.g. `compression_config='member.csv'`

**Comment by raayu83** [comment]
Hello,

just wanted to ask if there is any update on this topic?

Best regards

**Comment by YoannT** [comment]
Hi, I would also like to know if this feature will be added some day

**Comment by orlp** [comment]
It would be cool if Polars could read from .zip files directly as if it's part of a path, but it very quickly becomes rather hairy. Instead of just resolving paths we'd have to rewrite a bunch of code to also support a zip file being part of the path. And then you get questions about nested zip files (`my/foo.zip/bar.zip/test.csv`), zip files hosted on the network (`https://example.com/cool-collection.zip/test.parquet`), wildcards within zip files (`my/data.zip/2025-*.csv`), etc... Each of these needs custom handling for when it's a zip file rather than a folder, especially because I don't think anyone would find the naive solution of "fully unpack the zip file then continue" acceptable.

I'm not sure we're quite ready to support this.

**Comment by raayu83** [comment]
Don't these problems exist also for the existing compressed file formats supported by polars? (maybe apart from the first one?)

**Comment by orlp** [comment]
@raayu83 No, because the supported compression formats (gz, zstd, etc) are 1:1 with files, and aren't archives like zip's are.

**Comment by raayu83** [comment]
I see.
So basically while you could support a subset of zip compression that relies on only a single file being inside the archive rather easily, this would cause too much confusion about why zip files containing more than one file are not supported.

**Comment by mdavis-xyz** [comment]
I really want this feature. So if it's not going to be added to Polars because it complicates the API, would it be feasible to extend rust in a second library (e.g. maintained by me) dedicated to handling zips?

e.g. you can do:

```
with ZipFile(f, 'r') as z:
    with z.open('something.csv', 'r') as f:
        lf = pl.scan_csv(f)
```

That's slow, because the dataplane goes through Python not just Rust.

I don't yet know Rust. Is it feasible to rewrite this approach as a rust library that hooks into Polars via the Rust API, and then package that all up as a Python library?

**Comment by YoannT** [comment]
Couldn't you just limit the use case to 1:1 zip files and throw an error whenever that requirement is not met ?

It would eliminate any considerations about nested zipfiles etc.

Pandas does it like that I think.

(sorry if that was part of your previous answer but I don't know about the implications of zip file paths)
